Add TEE_ALG_RSASSA_PKCS1_V1_5

EDVTAZ opened the pull request:
I've been working with @etienne-lms, to add asymmetrical algorithms to the
SKS proposal. While testing with OpenSSH, I came across the usage of the
CKM_RSA_PKCS mechanism with the C_Sign function, which I couldn't implement
with the current cryptographic TEE API.

This extension allows signing and verifying raw/pre-hashed data,
without specifying a hash function. It is needed for implementing
the pkcs#11 CKM_RSA_PKCS mechanism for signing and verifying in SKS.
This patch adds the extension to the API and updates parts of
libtomcrypt to a newer version, to support this algorithm.

Signed-off-by: Gabor Szekely <szvgabor@gmail.com>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented:
This PR consists of two parts, one update of LTC and one update of OP-TEE.
Please split into two commits.
How about upstreaming to LTC, is that started?
Do you have a link to a PR?

EDVTAZ commented:
Ok, I will split it into two commits. 
Regarding libtomcrypt, I just took their latest version, and copied the necessary files into optee. The only modification I made had to do with some optimisation that was added some time back in optee to ltc. I modified the copied files, so they were in line with those.

jenswi-linaro commented:
OK, good. :-)

jforissier commented:
Yes good indeed, it means we should take as a separate (and prerequisite) task to update LTC to the latest version...

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -19,7 +19,7 @@
 #include <util.h>
 #if defined(CFG_CRYPTO_HKDF) || defined(CFG_CRYPTO_CONCAT_KDF) || \
-	defined(CFG_CRYPTO_PBKDF2)
+	defined(CFG_CRYPTO_PBKDF2) || defined(CFG_SECURE_KEY_STORAGE)
```
I would prefer another `CFG_`. Maybe introduce `CFG_CRYPTO_RSASSA` to default enable in the related make script.
I even wonder if we could not simply drop the condition and always include **tee_api_defines_extensions.h**.

etienne-lms commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3277,6 +3277,9 @@ TEE_Result syscall_asymm_operate(unsigned long state,
 		}
 		break;
 
+#if defined(CFG_SECURE_KEY_SERVICES)
```
I would prefer another `CFG_`. Maybe introduce `CFG_CRYPTO_RSASSA` to default enable in the related make script.

jenswi-linaro commented on core/lib/libtomcrypt/src/pk/rsa/rsa_sign_hash.c.bak:
```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: BSD-2-Clause
```
Please remove this file

etienne-lms commented:
Hi @EDVTAZ and thank for this PR. I think it is clean. Please try to format the commit comments as those already merged in the optee_os git tree. It helps maintenance.

@jforissier I have had a quick look a ltc update. Most changes are ok but few descriptors which requires updates in the **tee_ltc_provider.c**. Maybe there are other things to take care of..
The proposed changes here only syncs op-tee with libtomcrypt for this very `LTC_PKCS_1_V1_5_NA1` support. Would it be ok to merge it before a libtomcrypt full resync?

jforissier commented:
> judging by the timestamps in some of the files it's as old as 2007

No, the last global sync was against [a50cb361](https://github.com/libtom/libtomcrypt/commit/6ad52252688bb34f90b5e79da4830a927e87b81f) 2 years ago, and we have picked a few upstream commits since then.

jforissier commented:
@etienne-lms yes, we can merge this feature without a full LTC sync.

jforissier commented:
...and then we also want:
- a new test added to `xtest`
- some documentation in `documentation/extensions`.

EDVTAZ commented:
I removed the accidentally committed backup file, and some of the `#if`s. Also renamed `CFG_SECURE_KEY_SERVICES` to `CFG_CRYPTO_RSASSA_NA1`. I split the first commit into two and tried to rewrite the commit messages to be better.

Tomorrow I will investigate the IBART error and start working on an xtest test case and some documentation.

EDVTAZ commented:
IBART error was present, because I was working with an older version of OPTEE, that didn't have this commit: https://github.com/OP-TEE/optee_os/commit/e77020396508fc086d7a4d6137388b116e4a662f . I rebased my branch, so the tests should be good now.

EDVTAZ commented:
I created the xtest test case, see: https://github.com/OP-TEE/optee_test/pull/309

I also found some some errors in my code while testing, which I corrected in https://github.com/OP-TEE/optee_os/pull/2524/commits/06f2d218e894570b62ac401a0f5f22e6d8a517d8

EDVTAZ commented:
I have created the documentation for the extension, please let me know if it's all right.

jenswi-linaro commented on lib/libutee/include/tee_api_defines_extensions.h:
```diff
@@ -56,6 +56,12 @@
+ * PKCS#1 v1.5 RSASSA pre-hashed sign/verify
+ */
+
+#define TEE_ALG_RSASSA_PKCS1_V1_5	0x70007830
```
This defines a value in a range that as far as I can tell is reserved for future specifications.

jforissier replied:
@jenswi-linaro indeed this is a problem. The future Internal Core API specification v1.2 (currently numbered v1.1.2.51) gets rid of the structure of algorithm identifiers (table 6-12) and defines a whole range for implementation-defined identifiers (0xF0000000-0xF0FFFFFF). A value in that range would be an option I guess, but that certainly means more invasive changes to the OP-TEE code which currently assumes the structure defined in the current spec.

EDVTAZ replied:
@jenswi-linaro @jforissier then would `0xF0000830` be all right for this define?

jforissier replied:
@EDVTAZ I think it should be OK.

jforissier replied:
@EDVTAZ ...but `0xF0000001` would be even better.

EDVTAZ replied:
I've tried to make this work, but a lot of macros that are used all over the place rely on the identifier's value having the above mentioned structure. Most of the time the macros are called in switch statements, so I can't insert cases for an identifier that doesn't follow the structure. I've inserted if-else statements before switch statements to get around this, but it seems I'm missing something somewhere, because it fails to work. The resulting code is also ugly and hard to read.

Could we maybe find a value that works with the current macros, or just keep the current value until the changes @jforissier mentioned get implemented?

jbech-linaro replied:
@jforissier @EDVTAZ  , just a gentle reminder about this, since it seems like the discussion suddenly stopped (a since I also think it is a useful addition).

EDVTAZ commented:
Sorry this took so long, I've had a lot to do in the past month. The above two commits change the `TEE_ALG_RSASSA_PKCS1_V1_5` identifier from `0x70007830` to `0xF0000830`, so it is in the range of implementation-defined identifiers. This is the cleanest way I could come up with, to solve this problem without changing how `TEE_ALG_GET_CLASS` works. Please let me know if it's ok like this and if yes I will adjust the documentation to hold the right value as well.

jenswi-linaro commented on core/tee/tee_svc_cryp.c:
```diff
@@ -3428,15 +3441,22 @@ TEE_Result syscall_asymm_verify(unsigned long state,
-		if (data_len != hash_size) {
-			res = TEE_ERROR_BAD_PARAMETERS;
-			break;
+#ifdef CFG_CRYPTO_RSASSA_NA1
```
This ifdef and the one below clutters the code. It seems it would be harmless to remove them. If you agree please remove them, else we'll need to find some other way to avoid having them here.

EDVTAZ replied:
If I'm not mistaken, `TEE_ALG_RSASSA_PKCS1_V1_5` is included [here](https://github.com/EDVTAZ/optee_os/blob/4f0de9c225dca7a16b15aa78238659b254875108/core/tee/tee_svc_cryp.c#L23) in this file, so to remove the ifdef in question, we would need to remove all the ifdefs that guard the include that I linked.

jenswi-linaro replied:
This ifdeffing isn't OK so it will need to be fixed. I'm OK with unconditionally including `tee_api_defines_extensions.h`, another option is to put in a helper function `cs->algo != TEE_ALG_RSASSA_PKCS1_V1_5` that would return true if `CFG_CRYPTO_RSASSA_NA1` isn't defined. However, I'd prefer the less complicated solution.

EDVTAZ replied:
I removed the guards

EDVTAZ commented on core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:
```diff
@@ -79,11 +79,12 @@ int rsa_verify_hash_ex(const unsigned char *sig,      unsigned long siglen,
   }
 
-  if (padding == LTC_PKCS_1_PSS) {
+  if (padding != LTC_PKCS_1_V1_5_NA1) {
```
These changes were taken from the libtomcrypt repo: https://github.com/libtom/libtomcrypt/commit/aa4bae5ae9a278776b27fe817a988702ab0e9e3f#diff-f8abc0f8fcc24d9959cf4792e56ccce5L58 I think it is intentional, although it is not mentioned in the commit message.

EDVTAZ commented on core/lib/libtomcrypt/src/pk/rsa/rsa_verify_hash.c:
```diff
@@ -150,41 +144,62 @@ int rsa_verify_hash_ex(const unsigned char *sig,      unsigned long siglen,
+      /* test OID */
+      if ((reallen == outlen) &&
+          (digestinfo[0].size == hash_descriptor[hash_idx]->OIDlen) &&
+        (XMEMCMP(digestinfo[0].data, hash_descriptor[hash_idx]->OID, sizeof(unsigned long) * hash_descriptor[hash_idx]->OIDlen) == 0) &&
```
This change was introduced here: https://github.com/libtom/libtomcrypt/commit/e9c90e7f63eb25bb6751fd96cfdd074ff1710af8#diff-f8abc0f8fcc24d9959cf4792e56ccce5

EDVTAZ commented:
I've rebased and squashed the commits that belong together, I hope it's all right like this.  
I can't tell why the travis-ci fails, xtest passes locally.

jbech-linaro commented:
@EDVTAZ , Travis failed because it hit the time limit. I don't think we have to bother about that. IBART passes which indicates that things are fine.

jenswi-linaro commented:
For "￼libtomcrypt: port LTC_PKCS_1_V1_5_NA1 from ltc v1.18.2" please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`
For the others:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
Look ok to me, just a late comment:
`CFG_CRYPTO_RSASSA_NA1` is introduced in 2nd change ("core: crypto: add TEE_ALG_RSASSA_PKCS1_V1_5") but described in the comment of the 3rd change ("utee: support prehashed RSA sign/ver without ASN.1"). This is not a big issue.
Yet, a comment related to `CFG_CRYPTO_RSASSA_NA1`: can we add a change to default enable the feature (i.e `CFG_CRYPTO_RSASSA_NA1=y` in **mk/config.mk**). I see no reason to no support this extension in the default configuration.

EDVTAZ commented:
I have added `CFG_CRYPTO_RSASSA_NA1` to the default config.

jforissier commented:
For "mk/config.mk: default enable CFG_CRYPTO_RSASSA_NA1":
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
